### PR TITLE
+Add and use optional unscale args to reproducing_sum

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
This PR consists of a series of 4 commits that adds optional `unscale` arguments to the publicly visible `reproducing_sum()` interfaces in `MOM_coms`, and then uses them to refactor the routines in `MOM_sum_output` and 4 other modules to eliminate about 19 dimensional rescaling factors, and to work more broadly in rescaled variables for dimensional consistency testing.

  All answers are bitwise identical, but there are new optional arguments to 2 publicly visible interfaces and a new element in the transparent unit_scale_type.  This commits in this PR include:

- NOAA-GFDL/MOM6@193c83ec9 Use reproducing_sum with unscale in 5 places
- NOAA-GFDL/MOM6@842bbf4da Work in rescaled units in write_energy
- NOAA-GFDL/MOM6@4d255baf7 +Add RZL2_to_kg element to unit_scale_type
- NOAA-GFDL/MOM6@9e7e16523 +Add optional unscale argument to reproducing_sum

### CODE REVIEW

**Comment by marshallward** [comment]
1. Since the `unscale * array(..)` arguments also implicitly create scaled arrays as copies, you might be able to simplify the `reproducible_EFP_sum_*` functions with a pointer which conditionally points to these rescaled values for the same amount of work + memory usage.

   Something like
   ```
   real, intent(in), target :: array(..)
   real, pointer :: summed_array(..)

   if (present(unscale)) then
     allocate(summed_array, mold=array)
     summed_array(..) = unscale * array(..)
   else
     summed_array => array
   endif
   ```
   then eliminate `unscale_` and all of the conditional logic in the solver.

2. `field_chksum` is being redirected to `rotated_field_chksum` even though no `turns` argument is provided.  `rotated_field_chksum` is also redefined as some kind of generic rescaling/rotating wrapper for all debugging operations.  I am not quite comfortable with globbing these operations together.  It might be more appropriate to create a separate wrapper supporting `unscale`, or maybe just do it in-place?

    (`rotated_field_chksum` appears to only be used in one specific place of `MOM_restart.F90`.  Perhaps this function is not even necessary?)

**Comment by Hallberg-NOAA** [comment]
I have updated this PR to eliminate the changes to the `rotated_field_checksum()` interface.  I plan to revisit this later in a separate PR, probably with a different form.  This revised version also avoids using an internal variable name ending in an underscore, and it no longer contains any array-syntax math on multidimensional arrays.

File: src/diagnostics/MOM_sum_output.F90 [hunk 1]
```diff
@@ -532,7 +530,7 @@ subroutine write_energy(u, v, h, tv, day, n, G, GV, US, CS, tracer_CSp, dt_forci
   Isq = G%IscB ; Ieq = G%IecB ; Jsq = G%JscB ; Jeq = G%JecB
   isr = is - (G%isd-1) ; ier = ie - (G%isd-1) ; jsr = js - (G%jsd-1) ; jer = je - (G%jsd-1)
 
-  HL2_to_kg = GV%H_to_kg_m2*US%L_to_m**2
+  RZL4_to_J = US%RZL2_to_kg*US%L_T_to_m_s**2 ! Unused to unscale energies
```
Comment (marshallward) [inline src/diagnostics/MOM_sum_output.F90:533]: "Unused" -> "Used" ?

File: src/diagnostics/MOM_sum_output.F90 [hunk 2]
```diff
@@ -544,28 +542,22 @@ subroutine write_energy(u, v, h, tv, day, n, G, GV, US, CS, tracer_CSp, dt_forci
...
+  tmp1(:,:,:) = 0.0
+  do k=1,nz ; do j=js,je ; do i=is,ie
+    tmp1(i,j,k) = h(i,j,k) * (GV%H_to_RZ*areaTm(i,j))
+  enddo ; enddo ; enddo
+  mass_tot_RZL2 = reproducing_sum(tmp1, isr, ier, jsr, jer, sums=mass_lay, EFP_sum=mass_EFP, unscale=US%RZL2_to_kg)
 
-    mass_tot = reproducing_sum(tmp1, isr, ier, jsr, jer, sums=mass_lay, EFP_sum=mass_EFP)
-    do k=1,nz ; vol_lay(k) = (US%m_to_L**2*GV%H_to_Z/GV%H_to_kg_m2)*mass_lay(k) ; enddo
+  if (GV%Boussinesq) then
+    ! do k=1,nz ; vol_lay(k) = (GV%H_to_Z / GV%H_to_RZ) * mass_lay(k) ; enddo
```
Comment (marshallward) [inline src/diagnostics/MOM_sum_output.F90:552]: Keep this comment or remove?  (AFAIK R=GV%Rho0 in Boussinesq...)

File: src/diagnostics/MOM_sum_output.F90 (hunk 1 above)
Comment (Hallberg-NOAA) [reply src/diagnostics/MOM_sum_output.F90:533]: Corrected as suggested.  Thanks.

File: src/diagnostics/MOM_sum_output.F90 (hunk 2 above)
Comment (Hallberg-NOAA) [reply src/diagnostics/MOM_sum_output.F90:552]: This was removed as suggested.  Thanks.

**Review by marshallward — APPROVED** [review]
This looks very good now.  The change to `reproducing_sum_2d` is much better than the one I proposed, and no such change was needed to `reproducing_EFP_sum_3d`.  (My suggestion would have probably made it worse!)

**Comment by Hallberg-NOAA** [comment]
This PR has been tested at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/25745.

**Pull request merged**